### PR TITLE
Add command line interface

### PR DESCRIPTION
**Pull request by LiamPattinson**
In a recent chat, it was mentioned that a Command Line Interface (CLI) for basic operations would be helpful, and that it would serve as a useful first step towards a GUI. I've added a system using [`argparse`](https://docs.python.org/3/library/argparse.html) that allows users to convert GK input files on the command line, and they can optionally pull in equilibrium/kinetics/template data too.

As Pyrokinetics is capable of doing a bunch of different things, I designed this to be extendable using 'subparsers'. Currently, it only has one subcommand `convert`, but I could foresee it having subcommands such as `run`, `scan`, `compare`, `analyse`, etc in future.

Let me know if there's anything here you'd like me to change.

## Usage

You can run the CLI via two methods:

- `$ python3 -m pyrokinetics`, which uses the file `__main__.py`.
- `$ pyro`, which uses script entrypoints in `pyproject.toml` to call `pyrokinetics.cli.entrypoint:entrypoint` directly.

Both point to the same entrypoint function, so there's no practical difference in usage. If you only call `$ pyro`, it'll print a help message:

```
usage: Pyrokinetics [-h] {convert} ...

A tool to run and analyse gyrokinetics simulations.

optional arguments:
  -h, --help  show this help message and exit

subcommands:
  Please provide one of the following subcommands as a positional argument. For information on how each subcommand works, try providing '--help' after the subcommand.

  {convert}
    convert   Convert a gyrokinetics input file to a different code.
```

You can also get this by calling `$ pyro --help` or `$ pyro -h`. The `convert` subcommand also provides help text if you call `$ pyro convert --help`:

```
usage: Pyrokinetics convert [-h] [--input_type INPUT_TYPE] [--equilibrium EQUILIBRIUM] [--equilibrium_type EQUILIBRIUM_TYPE] [--kinetics KINETICS] [--kinetics_type KINETICS_TYPE] [--psi PSI] [--a_minor A_MINOR] [--output OUTPUT] [--template TEMPLATE] target input_file

positional arguments:
  target                The target gyrokinetics code, e.g. 'GS2', 'CGYRO'.
  input_file            The gyrokinetics config file you wish to convert.

optional arguments:
  -h, --help            show this help message and exit
  --input_type INPUT_TYPE
                        The code type of the input file. If not provided, this will be inferred.
  --equilibrium EQUILIBRIUM, --eq EQUILIBRIUM, -e EQUILIBRIUM
                        Path to a plasma equilibrium file, which is used to overwrite the flux surface in 'input_file'. Users should also provide 'psi' to select which flux surface to use from the equilibrium.
  --equilibrium_type EQUILIBRIUM_TYPE, --eq_type EQUILIBRIUM_TYPE
                        The type of equilibrium file. If not provided, this is inferred.
  --kinetics KINETICS, -k KINETICS
                        Path to a plasma kinetics file, which is used to overwrite the local species data in 'input_file'. Users should also provide 'psi' and 'a_minor' to select which flux surface to use, or provide 'psi' and 'equilibrium'.
  --kinetics_type KINETICS_TYPE, --k_type KINETICS_TYPE
                        The type of kinetics file. If not provided, this is inferred.
  --psi PSI, -p PSI     The normalised poloidal flux function, used to index which flux surface to draw equilibrium/kinetics data from. Should be in the range [0,1], with 0 being the magnetic axis, and 1 being the last closed flux surface.
  --a_minor A_MINOR, -a A_MINOR
                        The width of the last closed flux surface, in meters. Used to select a flux surface when providing kinetics data but no equilibrium. Otherwise, this argument is ignored.
  --output OUTPUT, -o OUTPUT
                        Name of the new gyrokinetics config file.
  --template TEMPLATE, -t TEMPLATE
                        Template file to use for the new gyrokinetics config file.
```

The following will convert a GS2 file to CGYRO:

```
$ pyro convert CGYRO my_gs2_file.in --output input.cgyro
```

You can also bring in equilibrium/kinetics stuff, but the command line gets a bit crowded if you do:

```
$ pyro convert CGYRO my_gs2_file.in --equilibrium my_eq.geqdsk --kinetics my_kinetics.cdf --psi=0.95 --output input.cgyro
```

Most optional arguments also have shortened aliases:

```
$ pyro convert CGYRO my_gs2_file.in -e my_eq.geqdsk -k my_kinetics.cdf -p=0.95 -o input.cgyro
```

## Issues

- Haven't written tests yet, and CLI stuff is quite difficult to test anyway. @ZedThree would you recommend just testing the `main()` function I've written and leaving the other pieces untested? Or should I test the whole thing using `subprocess`?
- Currently don't have a way to specify the preferred `LocalGeometry` type, but as that whole system is in flux I think it'll be fine to keep this restricted to Miller for now.

### CODE REVIEW

**Comment by ZedThree** [comment]
You can monkeypatch `argv` and then run the CLI in the unit tests, without the expense of using `subprocess`. See how we do it in Ford: https://github.com/Fortran-FOSS-Programmers/ford/blob/abe9451997b0920d5338f9129741dae7ccbade7d/test/test_example.py#L24-L27

```py
    with pytest.MonkeyPatch.context() as m:
        os.chdir(tmp_path)
        m.setattr(sys, "argv", ["ford", "example-project-file.md"])
        ford.run()
```

**Comment by bpatel2107** [comment]
I think #167 will mean that specifying the input and output geometries are essential! If the input isn't a standard `LocalGeometryMiller` (what will be the new one without squareness) then it won't be possible to directly convert between codes.

Might be worth waiting until that is merged, (which is ready to go!). The actual conversion is straight forward using `pyro.switch_local_geometry`. You would probably only need to specify the output geometry as this method doesn't need the input one.

**Comment by LiamPattinson** [comment]
@ZedThree Cheers for that, I'll take a look at how Ford handles things.

@bpatel2107 Good to know, I'll be sure to get a geometry option in there before finishing up. It should as simple as adding an argument `--geometry`/`-g` that takes something like 'Miller' or 'MXH' or 'Fourier' etc.

**Comment by bpatel2107** [comment]
Would also be good to have some basic plotting scripts for eigenfunctions, fluxes and growth rates, something like a `pyro_plot` script

**Comment by LiamPattinson** [comment]
I've added a test for the CLI 'convert' function. I first tried testing over all combinations of `gk_input`, `gk_output`, using equilibrium, using kinetics, switching geometry, using long option names, and explicitly listing types, but this resulted in over 500 individual tests and took about 10 minutes to run. I settled on 8 tests instead. Any ideas for how I could rewrite things to test a wider range of inputs/outputs without having the tests take forever?

I also tried rebasing the branch, and to avoid a messy merge conflict I figured I could get away with a force push. This seems to have broken the black CI (lesson learned, let's see how long it sticks this time). Should I close this PR and start a new one?

**Comment by bpatel2107** [comment]
Not sure about black. Maybe mark as ready for review and push an empty commit?

500 tests would be a tad excessive. I think as long as each component is covered by the test we should be fine, so maybe for each different option you can try for all the given options, i.e. try all the different equilibrium types, but don't necessarily run that for each gk code/kinetics type etc. So not an outer product.

**Comment by LiamPattinson** [comment]
I've rewritten the tests to cover all combinations of inputs, and the tests for equilibria, kinetics, etc only cover a single case. I'd like to leave this PR here, as now that the framework is ready it should be relatively easy for others to extend it to include additional subcommands. I agree that a plotting tool would be helpful, but I think it'd be best to leave to somebody who actually does gyrokinetics analysis and knows the best way to present that data.

**Comment by LiamPattinson** [comment]
(Please don't merge until the black thing is sorted, I don't know if it'll break things for the whole project)

**Comment by LiamPattinson** [comment]
Just going to try closing and reopening to see if that helps...

**Comment by LiamPattinson** [comment]
@ZedThree I appear to have broken things quite thoroughly. Any ideas?

**Comment by ZedThree** [comment]
I think it was most likely this commit: d2ebc51e85e45c14994ef087b3fcc160f50fad26
It runs `black` on PRs, which isn't necessary -- here it's trying to pull this branch from `origin`, i.e. this repo.

So `black.yml` should have `on: pull_request` removed. Might also be worth you pushing directly to this repo instead of a branch?

**Comment by bpatel2107** [comment]
I found that without that, black wouldn't run again after doing its automatic fix so I wasn't able to merge as the latest commit needs to pass the black test before merging

**Comment by LiamPattinson** [comment]
Whoops, only just read your last comment. Reverted my change.

**Comment by bpatel2107** [comment]
There may be a better way than what I did so happy to change if something else is preferable

**Comment by LiamPattinson** [comment]
Okay, I've now pushed it to pyrokinetics as well as PlasmaFAIR, and that seems to have fixed it. I guess I'll have to stop working from a fork in future. Thanks @ZedThree, and sorry for the notification spam.

File: pyrokinetics/cli/convert.py
Comment (bpatel2107) [inline pyrokinetics/cli/convert.py]: Technically there are a lot more now though I'm not sure if this is best place to put it but we have the following that are fully working
Miller (all)
MillerTurnbull (gene)
MXH (cgyro, tglf)

The fourier ones haven't been implemented in the `gk_code`, we can do the fit but haven't mapped it to the codes yet.

**Review by bpatel2107 — APPROVED** [review]
LGTM.

Only suggestion would be to list all the potential options for each type in the description. Maybe we can use `pyro.supported_gk_inputs` for example to list out the available `gk_code`?

File: pyrokinetics/cli/convert.py
Comment (LiamPattinson) [reply pyrokinetics/cli/convert.py]: Good catch, I'd previously implemented it to interpret 'Miller' and 'Fourier' differently depending on the target code, but since both CGYRO and GENE can handle plain Miller as well as their specialised variants, I couldn't figure out a clean way to implement it and fell back on just letting the user specify whatever they wanted. Looks like I forgot to rewrite the message to match.

**Review by bpatel2107 — APPROVED** [review]

**Comment by ZedThree** [comment]
> I found that without that, black wouldn't run again after doing its automatic fix so I wasn't able to merge as the latest commit needs to pass the black test before merging

Ok, another option might be to run the auto-black CI on `push` to `unstable` only. Then it won't stop PRs, but will always get applied

**Pull request merged**
